File: playwright-java/src/test/java/testtracks/steps/restfulbooker/api/AuthApiSteps.java
package testtracks.steps.restfulbooker.api;

import com.fasterxml.jackson.databind.node.ObjectNode;
import com.microsoft.playwright.APIResponse;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import org.junit.jupiter.api.Assertions;
import testtracks.steps.restfulbooker.api.auth.AuthApi;
import testtracks.steps.restfulbooker.api.auth.AuthToken;
import testtracks.support.ScenarioState;

public final class AuthApiSteps {
  private static final String AUTH_API_TARGET = "restfulBookerApi";

  private final ScenarioState state;
  private AuthToken authToken;
  private APIResponse authResponse;

  public AuthApiSteps(ScenarioState state) {
    this.state = state;
  }

  @When("valid admin credentials are submitted to the auth API")
  public void validAdminCredentialsAreSubmittedToTheAuthApi() {
    var data =
        state
            .data()
            .load(
                "api-auth", "scenarios.validAdminCredentialsProduceAToken", AuthApiTestData.class);

    var authApi = state.useApi(AUTH_API_TARGET, AuthApi::new);
    authToken = authApi.logIn(data.payload());
  }

  @Then("a reusable auth token is returned")
  public void aReusableAuthTokenIsReturned() {
    Assertions.assertNotNull(authToken);
    Assertions.assertFalse(authToken.token().isBlank());

    var authApi = state.useApi(AUTH_API_TARGET, AuthApi::new);
    var validation = authApi.validate(authToken.token());
    Assertions.assertTrue(validation.valid());
  }

  @When("invalid admin credentials are submitted to the auth API")
  public void invalidAdminCredentialsAreSubmittedToTheAuthApi() {
    var data =
        state
            .data()
            .load(
                "api-auth", "scenarios.invalidAdminCredentialsAreRejected", AuthApiTestData.class);

    var authApi = state.useApi(AUTH_API_TARGET, AuthApi::new);
    authResponse = authApi.tryLogIn(data.payload());
  }

  @Then("the credentials are rejected")
  public void theCredentialsAreRejected() {
    Assertions.assertNotNull(authResponse);
    Assertions.assertFalse(authResponse.ok());
  }

  public record AuthApiTestData(ObjectNode payload) {}
}

File: playwright-csharp/tests/Steps/RestfulBooker/API/AuthApiSteps.cs
using System.Text.Json.Nodes;
using Microsoft.Playwright;
using NUnit.Framework;
using Reqnroll;
using TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API.Auth;
using TestTracks.Playwright.CSharp.Specs.Support;

namespace TestTracks.Playwright.CSharp.Specs.Steps.RestfulBooker.API;

[Binding]
public sealed class AuthApiSteps
{
    private const string AuthApiTarget = "restfulBookerApi";

    private readonly ScenarioState _state;
    private AuthToken? _authToken;
    private IAPIResponse? _authResponse;

    public AuthApiSteps(ScenarioState state)
    {
        _state = state;
    }

    [When("valid admin credentials are submitted to the auth API")]
    public async Task WhenValidAdminCredentialsAreSubmittedToTheAuthApi()
    {
        var data = _state.Data.Load<ValidAuthApiTestData>(
            "api-auth",
            "scenarios.validAdminCredentialsProduceAToken");

        var authApi = await _state.UseApiAsync(
            AuthApiTarget,
            (settings, request, evidence) => new AuthApi(settings, request, evidence));

        _authToken = await authApi.LogIn(data.Payload);
    }

    [Then("a reusable auth token is returned")]
    public async Task ThenAReusableAuthTokenIsReturned()
    {
        Assert.That(_authToken?.Token, Is.Not.Null.And.Not.Empty);
        var authApi = await _state.UseApiAsync(
            AuthApiTarget,
            (settings, request, evidence) => new AuthApi(settings, request, evidence));

        var validation = await authApi.Validate(_authToken!.Token);
        Assert.That(validation.Valid, Is.True);
    }

    [When("invalid admin credentials are submitted to the auth API")]
    public async Task WhenInvalidAdminCredentialsAreSubmittedToTheAuthApi()
    {
        var data = _state.Data.Load<InvalidAuthApiTestData>(
            "api-auth",
            "scenarios.invalidAdminCredentialsAreRejected");

        var authApi = await _state.UseApiAsync(
            AuthApiTarget,
            (settings, request, evidence) => new AuthApi(settings, request, evidence));

        _authResponse = await authApi.TryLogIn(data.Payload);
    }

    [Then("the credentials are rejected")]
    public void ThenTheCredentialsAreRejected()
    {
        Assert.That(_authResponse, Is.Not.Null);
        Assert.That(_authResponse!.Ok, Is.False);
    }
}

public sealed record ValidAuthApiTestData(JsonObject Payload);

public sealed record InvalidAuthApiTestData(JsonObject Payload);
